File: src/parse.ts
// f/vless/parse.ts

import type { ProxyRecord } from "./types";
import { PROXY_SCHEMES, PLACEHOLDER_HOSTS } from "./constants";

export function parseAllLines(allLines: string[]): {
  records: ProxyRecord[];
  skipped: number;
  duplicates: number;
} {
  const seen = new Set<string>();
  const records: ProxyRecord[] = [];
  let skipped = 0;
  let duplicates = 0;

  for (const line of allLines) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) {
      skipped++;
      continue;
    }

    if (!PROXY_SCHEMES.some((s) => trimmed.startsWith(s))) {
      skipped++;
      continue;
    }

    let record: ProxyRecord | null = null;

    if (trimmed.startsWith("vless://")) {
      record = parseVless(trimmed);
    } else if (trimmed.startsWith("vmess://")) {
      record = parseVmess(trimmed);
    } else if (trimmed.startsWith("trojan://")) {
      record = parseTrojan(trimmed);
    } else if (trimmed.startsWith("ss://")) {
      record = parseSs(trimmed);
    }

    if (!record) {
      skipped++;
      continue;
    }

    // Validate
    if (!record.host || record.port <= 0 || record.port > 65535) {
      skipped++;
      continue;
    }

    if (PLACEHOLDER_HOSTS.has(record.host)) {
      skipped++;
      continue;
    }

    // Dedup by (host, port, protocol)
    const dedupKey = `${record.host}:${record.port}:${record.protocol}`;
    if (seen.has(dedupKey)) {
      duplicates++;
      continue;
    }
    seen.add(dedupKey);

    records.push(record);
  }

  return { records, skipped, duplicates };
}

function parseVless(line: string): ProxyRecord | null {
  try {
    // Strip trailing garbage characters
    let cleaned = line.replace(/\\+$/, "").replace(/%22,%22/g, "").replace(/%22/g, "");
    // Fix double-escaped backslashes in query params
    cleaned = cleaned.replace(/\\\\+/g, "");

    const url = new URL(cleaned);
    const host = url.hostname;
    const port = parseInt(url.port, 10) || 443;
    const uuid = decodeURIComponent(url.username);
    const fragment = decodeURIComponent(url.hash.slice(1));
    const queryParams: Record<string, string> = {};
    url.searchParams.forEach((v, k) => {
      queryParams[k] = v;
    });

    // Normalize insecure flags
    const normalized = normalizeInsecure(queryParams);

    return {
      protocol: "vless",
      host,
      port,
      uuidOrPassword: uuid,
      queryParams: normalized,
      fragment,
      originalLine: line,
    };
  } catch {
    return null;
  }
}

function parseVmess(line: string): ProxyRecord | null {
  try {
    let encoded = line.slice("vmess://".length);
    // Normalize URL-safe base64
    encoded = encoded.replace(/-/g, "+").replace(/_/g, "/");
    while (encoded.length % 4 !== 0) {
      encoded += "=";
    }
    const decoded = Buffer.from(encoded, "base64").toString("utf-8");
    const config = JSON.parse(decoded);

    const host = config.add || config.address;
    const port = parseInt(config.port, 10);
    const uuid = config.id;
    const fragment = config.ps || "";

    const queryParams: Record<string, string> = {};
    if (config.net) queryParams.type = config.net;
    if (config.tls === "tls") queryParams.security = "tls";
    if (config.sni) queryParams.sni = config.sni;
    if (config.path) queryParams.path = config.path;
    if (config.host) queryParams.host = config.host;
    if (config.flow) queryParams.flow = config.flow;

    return {
      protocol: "vmess",
      host,
      port,
      uuidOrPassword: uuid,
      queryParams,
      fragment,
      originalLine: line,
    };
  } catch {
    return null;
  }
}

function parseTrojan(line: string): ProxyRecord | null {
  try {
    const url = new URL(line);
    const host = url.hostname;
    const port = parseInt(url.port, 10) || 443;
    const password = decodeURIComponent(url.username);
    const fragment = decodeURIComponent(url.hash.slice(1));
    const queryParams: Record<string, string> = {};
    url.searchParams.forEach((v, k) => {
      queryParams[k] = v;
    });

    const normalized = normalizeInsecure(queryParams);

    return {
      protocol: "trojan",
      host,
      port,
      uuidOrPassword: password,
      queryParams: normalized,
      fragment,
      originalLine: line,
    };
  } catch {
    return null;
  }
}

function parseSs(line: string): ProxyRecord | null {
  try {
    let encoded = line.slice("ss://".length);
    // ss://base64@host:port#fragment or ss://method:password@host:port#fragment
    const hashIdx = encoded.indexOf("#");
    let fragment = "";
    let main = encoded;
    if (hashIdx !== -1) {
      fragment = decodeURIComponent(encoded.slice(hashIdx + 1));
      main = encoded.slice(0, hashIdx);
    }

    // Try base64@host:port format
    const atIdx = main.lastIndexOf("@");
    if (atIdx === -1) {
      // Try full base64
      let b64 = main.replace(/-/g, "+").replace(/_/g, "/");
      while (b64.length % 4 !== 0) b64 += "=";
      const decoded = Buffer.from(b64, "base64").toString("utf-8");
      const innerAt = decoded.indexOf("@");
      if (innerAt === -1) return null;
      const methodPassword = decoded.slice(0, innerAt);
      const hostPort = decoded.slice(innerAt + 1);
      const colonIdx = methodPassword.indexOf(":");
      const method = methodPassword.slice(0, colonIdx);
      const password = methodPassword.slice(colonIdx + 1);
      const lastColon = hostPort.lastIndexOf(":");
      const host = hostPort.slice(0, lastColon);
      const port = parseInt(hostPort.slice(lastColon + 1), 10);

      return {
        protocol: "ss",
        host,
        port,
        uuidOrPassword: password,
        queryParams: { method },
        fragment,
        originalLine: line,
      };
    }

    const methodPassword = main.slice(0, atIdx);
    const hostPort = main.slice(atIdx + 1);
    const colonIdx = methodPassword.indexOf(":");
    const method = methodPassword.slice(0, colonIdx);
    const password = methodPassword.slice(colonIdx + 1);
    const lastColon = hostPort.lastIndexOf(":");
    const host = hostPort.slice(0, lastColon);
    const port = parseInt(hostPort.slice(lastColon + 1), 10);

    return {
      protocol: "ss",
      host,
      port,
      uuidOrPassword: password,
      queryParams: { method },
      fragment,
      originalLine: line,
    };
  } catch {
    return null;
  }
}

function normalizeInsecure(params: Record<string, string>): Record<string, string> {
  const result = { ...params };
  if (result.allowInsecure || result.insecure || result.allow_insecure) {
    result.insecure = "1";
    delete result.allowInsecure;
    delete result.allow_insecure;
  }
  return result;
}

export function applyNameFilter(
  records: ProxyRecord[],
  nameInclude: string,
  nameExclude: string
): ProxyRecord[] {
  let filtered = records;
  if (nameExclude) {
    const re = new RegExp(nameExclude);
    filtered = filtered.filter((r) => !re.test(r.fragment));
  }
  if (nameInclude) {
    const re = new RegExp(nameInclude);
    filtered = filtered.filter((r) => re.test(r.fragment));
  }
  return filtered;
}

File: internal/parse/parse.go
package parse

import (
	"encoding/base64"
	"encoding/json"
	"net/url"
	"strconv"
	"strings"

	"github.com/michael/vless-sub-server/internal/config"
)

var proxySchemes = []string{"vless://", "vmess://", "trojan://", "ss://"}

type ParseResult struct {
	Records   []ProxyRecord
	Skipped   int
	Duplicates int
}

func ParseAllLines(lines []string) ParseResult {
	seen := map[string]bool{}
	var records []ProxyRecord
	skipped, duplicates := 0, 0

	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			skipped++
			continue
		}

		matched := false
		for _, s := range proxySchemes {
			if strings.HasPrefix(trimmed, s) {
				matched = true
				break
			}
		}
		if !matched {
			skipped++
			continue
		}

		var record *ProxyRecord
		switch {
		case strings.HasPrefix(trimmed, "vless://"):
			record = parseVless(trimmed)
		case strings.HasPrefix(trimmed, "vmess://"):
			record = parseVMess(trimmed)
		case strings.HasPrefix(trimmed, "trojan://"):
			record = parseTrojan(trimmed)
		case strings.HasPrefix(trimmed, "ss://"):
			record = parseSS(trimmed)
		}

		if record == nil || record.Host == "" || record.Port <= 0 || record.Port > 65535 {
			skipped++
			continue
		}

		if config.PlaceholderHosts[record.Host] {
			skipped++
			continue
		}

		key := record.Host + ":" + strconv.Itoa(record.Port) + ":" + string(record.Protocol) + ":" + record.UUIDOrPassword
		if seen[key] {
			duplicates++
			continue
		}
		seen[key] = true
		records = append(records, *record)
	}

	return ParseResult{Records: records, Skipped: skipped, Duplicates: duplicates}
}

func ApplyNameFilter(records []ProxyRecord, include, exclude string) []ProxyRecord {
	filtered := records
	if exclude != "" {
		result := filtered[:0]
		for _, r := range filtered {
			if !strings.Contains(r.Fragment, exclude) {
				result = append(result, r)
			}
		}
		filtered = result
	}
	if include != "" {
		result := filtered[:0]
		for _, r := range filtered {
			if strings.Contains(r.Fragment, include) {
				result = append(result, r)
			}
		}
		filtered = result
	}
	return filtered
}

func parseVless(line string) *ProxyRecord {
	cleaned := strings.TrimRight(line, "\\")
	u, err := url.Parse(cleaned)
	if err != nil {
		return nil
	}
	port := 443
	if p := u.Port(); p != "" {
		if v, err := strconv.Atoi(p); err == nil {
			port = v
		}
	}
	params := map[string]string{}
	for k, v := range u.Query() {
		if len(v) > 0 {
			params[k] = v[0]
		} else {
			params[k] = ""
		}
	}
	normalizeInsecure(params)

	return &ProxyRecord{
		Protocol:       VLESS,
		Host:           u.Hostname(),
		Port:           port,
		UUIDOrPassword: u.User.Username(),
		QueryParams:    params,
		Fragment:       u.Fragment,
		OriginalLine:   line,
	}
}

func parseVMess(line string) *ProxyRecord {
	encoded := line[len("vmess://"):]
	encoded = strings.ReplaceAll(encoded, "-", "+")
	encoded = strings.ReplaceAll(encoded, "_", "/")
	for len(encoded)%4 != 0 {
		encoded += "="
	}
	decoded, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		return nil
	}

	var cfg struct {
		V             string `json:"v"`
		Host          string `json:"add"`
		Port          any    `json:"port"`
		ID            string `json:"id"`
		PS            string `json:"ps"`
		Net           string `json:"net"`
		Type          string `json:"type"`
		TLS           string `json:"tls"`
		SNI           string `json:"sni"`
		Path          string `json:"path"`
		Host2         string `json:"host"`
		Flow          string `json:"flow"`
		Scy           string `json:"scy"`
		Alpn          string `json:"alpn"`
		FP            string `json:"fp"`
		PBK           string `json:"pbk"`
		SID           string `json:"sid"`
		SPX           string `json:"spx"`
		AllowInsecure bool   `json:"allowInsecure"`
	}
	if err := json.Unmarshal(decoded, &cfg); err != nil {
		return nil
	}

	if cfg.V != "" && cfg.V != "2" {
		return nil
	}

	// Port can be int or string
	port := 0
	switch p := cfg.Port.(type) {
	case float64:
		port = int(p)
	case string:
		port, _ = strconv.Atoi(p)
	}

	params := map[string]string{}
	if cfg.Net != "" {
		params["type"] = cfg.Net
	}
	if cfg.TLS == "tls" {
		params["security"] = "tls"
	} else if cfg.TLS == "reality" {
		params["security"] = "reality"
	}
	if cfg.SNI != "" {
		params["sni"] = cfg.SNI
	}
	if cfg.Path != "" {
		params["path"] = cfg.Path
	}
	if cfg.Host2 != "" {
		params["host"] = cfg.Host2
	}
	if cfg.Flow != "" {
		params["flow"] = cfg.Flow
	}
	if cfg.Type != "" && cfg.Type != "none" {
		params["headerType"] = cfg.Type
	}
	if cfg.Scy != "" {
		params["scy"] = cfg.Scy
	}
	if cfg.Alpn != "" {
		params["alpn"] = cfg.Alpn
	}
	if cfg.FP != "" {
		params["fp"] = cfg.FP
	}
	if cfg.PBK != "" {
		params["pbk"] = cfg.PBK
	}
	if cfg.SID != "" {
		params["sid"] = cfg.SID
	}
	if cfg.SPX != "" {
		params["spx"] = cfg.SPX
	}
	if cfg.AllowInsecure {
		params["insecure"] = "1"
	}
	normalizeInsecure(params)

	return &ProxyRecord{
		Protocol:       VMess,
		Host:           cfg.Host,
		Port:           port,
		UUIDOrPassword: cfg.ID,
		QueryParams:    params,
		Fragment:       cfg.PS,
		OriginalLine:   line,
	}
}

func parseTrojan(line string) *ProxyRecord {
	u, err := url.Parse(line)
	if err != nil {
		return nil
	}
	if u.User == nil {
		return nil
	}
	port := 443
	if p := u.Port(); p != "" {
		if v, err := strconv.Atoi(p); err == nil {
			port = v
		}
	}
	password := u.User.Username()
	if p, ok := u.User.Password(); ok {
		password += ":" + p
	}
	params := map[string]string{}
	for k, v := range u.Query() {
		if len(v) > 0 {
			params[k] = v[0]
		} else {
			params[k] = ""
		}
	}
	normalizeInsecure(params)

	return &ProxyRecord{
		Protocol:       Trojan,
		Host:           u.Hostname(),
		Port:           port,
		UUIDOrPassword: password,
		QueryParams:    params,
		Fragment:       u.Fragment,
		OriginalLine:   line,
	}
}

func parseSS(line string) *ProxyRecord {
	encoded := line[len("ss://"):]
	hashIdx := strings.Index(encoded, "#")
	var fragment, main string
	if hashIdx != -1 {
		fragment = encoded[hashIdx+1:]
		main = encoded[:hashIdx]
	} else {
		main = encoded
	}

	atIdx := strings.LastIndex(main, "@")
	if atIdx == -1 {
		return nil
	}
	methodPassword := main[:atIdx]
	hostPort := main[atIdx+1:]

	colonIdx := strings.Index(methodPassword, ":")
	if colonIdx == -1 {
		return nil
	}
	method := methodPassword[:colonIdx]
	password := methodPassword[colonIdx+1:]

	lastColon := strings.LastIndex(hostPort, ":")
	if lastColon == -1 {
		return nil
	}
	host := hostPort[:lastColon]
	port, err := strconv.Atoi(hostPort[lastColon+1:])
	if err != nil {
		return nil
	}

	return &ProxyRecord{
		Protocol:       SS,
		Host:           host,
		Port:           port,
		UUIDOrPassword: password,
		QueryParams:    map[string]string{"method": method},
		Fragment:       fragment,
		OriginalLine:   line,
	}
}

func normalizeInsecure(params map[string]string) {
	val := params["allowInsecure"]
	if val == "" {
		val = params["insecure"]
	}
	if val == "" {
		val = params["allow_insecure"]
	}
	if val != "" {
		params["insecure"] = "1"
	}
	delete(params, "allowInsecure")
	delete(params, "allow_insecure")
}
